src/main/java/co/cask/spanner/source/SpannerRecordReader.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.spanner.source;
+
+import com.google.cloud.spanner.ResultSet;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * Spanner record reader - updates result set during iteration
+ */
+public class SpannerRecordReader extends RecordReader<NullWritable, ResultSet> {
+  private static final Logger LOG = LoggerFactory.getLogger(SpannerRecordReader.class);
+  PartitionInputSplit partitionInputSplit;
+  ResultSet resultSet;
+  boolean isFinished;
+
+  @Override
+  public void initialize(InputSplit inputSplit,
+                         TaskAttemptContext taskAttemptContext) throws IOException, InterruptedException {
+    partitionInputSplit = (PartitionInputSplit) inputSplit;
+    try {
+      resultSet = partitionInputSplit.getResultSet();
```

Comment:
you should only be getting the Partition from the input split. Reading logic should be moved from the split to the reader.

Reply:
thanks, that really simplifies the logic.